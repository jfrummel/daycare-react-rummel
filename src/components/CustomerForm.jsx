import React, { Component } from 'react';



class CustomerForm extends Component {
    constructor(props) {
        super(props);

        this.state = {
            firstName: props.customer ? props.customer.firstName : "",
            lastName: props.customer ? props.customer.lastName : "",
            numOfChildren: props.customer ? props.customer.numOfChildren : 0,
            hourlyRate: props.customer ? (props.customer.hourlyRate.toString()) : "",
            error: ""
        }
    }

    onFirstNameChange = (e) => {
        const firstName = e.target.value;
        this.setState(() => ({ firstName }));
    };

    onLastNameChange = (e) => {
        const lastName = e.target.value;
        this.setState(() => ({ lastName }));
    };

    onNumChildrenChange = (e) => {
        const numOfChildren = e.target.value;
        this.setState(() => ({ numOfChildren }));
    };

    onHourlyRateChange = (e) => {
        const hourlyRate = e.target.value;
        if (!hourlyRate || hourlyRate.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ hourlyRate }));
        }
    };

    onSubmit = (e) => {
        e.preventDefault();

        if (!this.state.firstName || !this.state.lastName) {
            this.setState(() => ({ error: "First and Last Name Required" }))
        } else {
            this.setState(() => ({ error: "" }));
            this.props.onSubmit({
                firstName: this.state.firstName,
                lastName: this.state.lastName,
                numOfChildren: this.state.numOfChildren,
                hourlyRate: parseFloat(this.state.hourlyRate)
            });
        };
    };

    render() {
        return (
            <form onSubmit={this.onSubmit}>
                {this.state.error && <p>{this.state.error}</p>}
                <input
                    type="text"
                    value={this.state.firstName}
                    placeholder="First Name"
                    onChange={this.onFirstNameChange}
                    autoFocus
                />
                <input
                    type="text"
                    value={this.state.lastName}
                    placeholder="Last Name"
                    onChange={this.onLastNameChange}
                />
                <input
                    type="number"
                    value={this.state.numOfChildren}
                    placeholder="Number of Children"
                    onChange={this.onNumChildrenChange}
                />
                <input
                    type="text"
                    value={this.state.hourlyRate}
                    placeholder="Hourly Rate"
                    onChange={this.onHourlyRateChange}
                />
                <button>{this.props.buttonText}</button>
            </form>
        );
    }
}

export default CustomerForm;