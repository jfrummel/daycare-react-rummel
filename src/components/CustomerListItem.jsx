import React from 'react';
import { Link } from 'react-router-dom';

const CustomerListItem = (props) => (
    <li>
        <Link to={`edit/${props.customer.id}`}>{props.customer.firstName} {props.customer.lastName}</Link>
        <p>Number of children: {props.customer.numOfChildren}</p>
        <p>Hourly rate: ${props.customer.hourlyRate}</p>
    </li>
);

export default CustomerListItem;