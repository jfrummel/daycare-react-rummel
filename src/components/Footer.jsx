import React from 'react';

const Footer = () => (
    <footer>
        <p>&copy; JR Designs 2020</p>
    </footer>
);

export default Footer;