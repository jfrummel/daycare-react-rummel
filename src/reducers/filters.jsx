// filters reducers

// Default state

const filtersDefaultState = {
    text: ''
};

const filtersReducer = (state = filtersDefaultState, action) => {
    switch (action.type) {
        case "SET_TEXT_FILTER":
            return {
                ...state, text: action.text
            }
        default:
            return state;
    }
};

export default filtersReducer;