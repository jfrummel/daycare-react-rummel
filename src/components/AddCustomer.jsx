import React from 'react';
import CustomerForm from './CustomerForm';
import { connect } from 'react-redux';
import { addCustomer } from '../actions/customers';

const AddCustomer = (props) => (
    <div>
        <h2>Add a new customer</h2>
        <CustomerForm
            buttonText="Add Customer"
            onSubmit={(customer) => {
                props.dispatch(addCustomer(customer));
                props.history.push("/customer_list")
            }}
        />
    </div>
);

export default connect()(AddCustomer);