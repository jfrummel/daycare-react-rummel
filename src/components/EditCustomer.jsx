import React from 'react';
import { connect } from 'react-redux';
import { editCustomer, removeCustomer } from '../actions/customers';
import CustomerForm from './CustomerForm';

const EditCustomer = (props) => (
    <div>
        <h3>Edit Customer</h3>
        <CustomerForm
            customer={props.customer}
            buttonText="Save Expense"
            onSubmit={(customer) => {
                props.dispatch(editCustomer(props.customer.id, customer));
                props.history.push("/customer_list");
            }}
        />
        <button
            onClick={() => {
                props.dispatch(removeCustomer({ id: props.customer.id }));
                props.history.push("/customer_list");
            }}
        >
            Remove
        </button>
    </div>
);

const mapStateToProps = (state, props) => {
    return {
        customer: state.customers.find((customer) => customer.id === props.match.params.id)
    }
};

export default connect(mapStateToProps)(EditCustomer);