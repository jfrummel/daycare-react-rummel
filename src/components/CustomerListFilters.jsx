import React, { Component } from 'react';
import { connect } from 'react-redux';
import { setTextFilter } from '../actions/filters';


export class CustomerListFilters extends Component {

    onTextChange = (e) => {
        const text = e.target.value;
        this.props.setTextFilter(text);
    };

    render() {
        return (
            <div>
                <input
                    type="text"
                    value={this.props.filters.text}
                    onChange={this.onTextChange}
                />
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text))
});

const mapStateToProps = (state) => ({
    filters: state.filters
});

export default connect(mapStateToProps, mapDispatchToProps)(CustomerListFilters);