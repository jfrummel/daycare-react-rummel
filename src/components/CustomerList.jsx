import React from 'react';
import { connect } from 'react-redux';
import { getVisibleCustomers } from '../selectors/customers';
import CustomerListFilters from './CustomerListFilters';
import CustomerListItem from './CustomerListItem';

export const CustomerList = (props) => (
    <div>
        <CustomerListFilters />
        {props.customers.length === 0 ? <h3>No Customers found</h3> :
            <ul>
                {props.customers.map((customer) => {
                    return <CustomerListItem key={customer.id} customer={customer} />
                })}
            </ul>
        }

    </div>
);

const mapStateToProps = (state) => ({
    customers: getVisibleCustomers(state.customers, state.filters)
});

export default connect(mapStateToProps)(CustomerList);