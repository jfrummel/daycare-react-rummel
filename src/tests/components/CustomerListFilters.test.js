import React from 'react';
import { shallow } from 'enzyme';
import { CustomerListFilters } from '../../components/CustomerListFilters';
import { filters, altFilters } from '../fixtures/filters';

let setTextFilter, wrapper;

beforeEach(() => {
    setTextFilter = jest.fn();
    wrapper = shallow(<CustomerListFilters filters={filters} setTextFilter={setTextFilter} />)
});

test('renders customer list filters component correctly', () => {
    expect(wrapper).toMatchSnapshot();
});

test('renders customer list filters component correctly with alt filters', () => {
    wrapper.setProps({ filters: altFilters });
    expect(wrapper).toMatchSnapshot();
});

test('handles set text filter correctly', () => {
    const value = "Lauren";
    wrapper.find('input').at(0).simulate('change', {
        target: { value }
    });
    expect(setTextFilter).toHaveBeenLastCalledWith(value);
    expect(wrapper).toMatchSnapshot();
});

