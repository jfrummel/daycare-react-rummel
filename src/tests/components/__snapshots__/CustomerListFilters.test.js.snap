// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handles set text filter correctly 1`] = `
<div>
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
</div>
`;

exports[`renders customer list filters component correctly 1`] = `
<div>
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
</div>
`;

exports[`renders customer list filters component correctly with alt filters 1`] = `
<div>
  <input
    onChange={[Function]}
    type="text"
    value="Trajan"
  />
</div>
`;
